mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/service/impl/MobileAcademyServiceImpl.java

```diff
@@ -51,99 +63,111 @@
-            DateTime version = course.getModificationDate();
-            return version.getMillis();
-        } else {
-            // return -1 and let the caller handle the upstream response
-            return -1;
+        NmsCourse course = nmsCourseDataService.getCourseByName(COURSE_NAME);
+
+        if (course == null) {
+            alertService.create("mTraining.Course", COURSE_NAME, "Could not find course", AlertType.CRITICAL, AlertStatus.NEW, 0, null);
+            throw new IllegalStateException("No course bootstrapped. Check deployment");
         }
+
+        return course.getModificationDate().getMillis();
     }
 
     @Override
     public MaBookmark getBookmark(Long callingNumber, Long callId) {
 
-        List<Bookmark> bookmarks = bookmarkDataService.findBookmarksForUser(callingNumber.toString());
-        if (CollectionUtils.isEmpty(bookmarks)) {
-            return null;
-        }
+        Bookmark existingBookmark = bookmarkService.getLatestBookmarkByUserId(callingNumber.toString());
 
-        if (bookmarks.size() > 1) {
-            LOGGER.debug("Found more than 1 instance of valid bookmark, picking top");
+        if (existingBookmark != null) {
+            MaBookmark toReturn = setMaBookmarkProperties(existingBookmark);
+            toReturn.setCallId(callId);
+            return toReturn;
         }
 
-        Bookmark existingBookmark = bookmarks.get(0);
-        MaBookmark toReturn = setMaBookmarkProperties(existingBookmark);
-        toReturn.setCallId(callId);
-        return  toReturn;
+        return null;
     }
 
     @Override
     public void setBookmark(MaBookmark saveBookmark) {
 
         String callingNumber = saveBookmark.getCallingNumber().toString();
-        List<Bookmark> existing = bookmarkDataService.findBookmarksForUser(callingNumber);
+        Bookmark existingBookmark = bookmarkService.getLatestBookmarkByUserId(callingNumber);
 
-        if (CollectionUtils.isEmpty(existing)) {
+        if (existingBookmark == null) {
             // if no bookmarks exist for user
             LOGGER.info("No bookmarks found for user " + callingNumber);
-            bookmarkDataService.create(setBookmarkProperties(saveBookmark, new Bookmark()));
+            bookmarkService.createBookmark(setBookmarkProperties(saveBookmark, new Bookmark()));
         } else {
-            // error check
-            if (existing.size() > 1) {
-                LOGGER.error("Found more than 1 bookmark for calling number. This should never be possible.");
-                LOGGER.error("Contact dev team about calling number: " + callingNumber);
-            }
 
             // update the first bookmark
-            LOGGER.info("Updating the first bookmark for user");
-            bookmarkDataService.update(setBookmarkProperties(saveBookmark, existing.get(0)));
+            LOGGER.info("Updating bookmark for user");
+            bookmarkService.updateBookmark(setBookmarkProperties(saveBookmark, existingBookmark));
```

Comment:
you are right. That would be a bad request exception that needs to be checked in the api controller as well. I'll update and push out another patch for this.